koitoror/SendIT
# Ft parcel orders endpoints 162030527

#### What does this PR do?
Creates all the required parcel delivery order API endpoints to : 
Performs LFA Feedback

#### Description of Task to be completed?
- Fetch all parcel delivery orders - for all users [by admin]
- Create a parcel delivery order [by user]
- Fetch a single parcel delivery order [by user]
- Delete a single parcel delivery order [by admin]
- Modify the present location and status of a parcel delivery order [by admin]
- Modify the status, to cancel, a delivery order [by user]
- Fetch all parcel delivery orders - for single user [by user]

####Feedback
- Remove commented code
- Test return messages
- Work on validation of endpoints
- Fix the "Null" issue
- Make sure admin routes are authorized correctly
- Add logout features
- Make descriptors more specific
- Have all tests working


#### How should this be manually tested?
For the Tests to work you need to do the following steps

Clone the repository into your local environment

```
$ git clone https://github.com/koitoror/SendIT.git
```

Switch to the PARCELS branch
```
$ git checkout ft-parcel-orders-endpoints-162030527
```

Navigate to api folder.
```
cd SendIT
```

Create a virtual environment.
Use [virtualenv](http://www.pythonforbeginners.com/basics/how-to-use-python-virtualenv) for an isolated working environment.

```
virtualenv venv --python=python3
```

Install the packages.
```
pip3 install -r requirements.txt
```

Set environment variables for 

> `SECRET_KEY` is your secret key

> `FLASK_CONFIG` is the environment you are running on. Should be either `Production`, `Development` or `Testing`. NOTE: its case sensitive

To get the app running...

```bash
$ python run.py run
```

Open the root path in your browser to test the endpoints. 
You can also use Postman or any other agent to test the endpoints

On Postman enter the relevant URL from below list which is

**EndPoint** | **Functionality**
--- | ---
GET  `/api/v2/parcels` | Fetch all parcel delivery orders for all users by admin
POST  `/api/v2/parcels` | Create a parcel delivery order by user
GET  `/api/v2/parcels/<parcelId>` | Fetch a single parcel delivery order by user
DELETE  `/api/v2/parcels/<parcelId>` | Delete a parcel delivery order by admin
PUT  `/api/v2/parcels/<parcelId>/cancel` | Cancel a parcel delivery order by user
PUT  `/api/v2/parcels/<parcelId>/destination` | Change the destination of a parcel delivery order by user
PUT  `/api/v2/parcels/<parcelId>/presentLocation` | Change the present location of a parcel delivery order by admin
PUT  `/api/v2/parcels/<parcelId>/status` | Change the status of a parcel delivery order by admin
GET  `/api/v2/users/<userId>/parcels` | Fetch all parcel delivery orders by user


#### Any background context you want to provide?
- Loaded the app runners, blueprints,  tests integrations, and configurations.
- Did chores of updating the Files, Directory Restructure and Code refactors

#### What are the relevant pivotal tracker stories?
[#162030527](https://www.pivotaltracker.com/n/projects/2216466)

#### Screenshots (if appropriate)
